src/main/java/tc/oc/pgm/map/NameCacheUtil.java

```diff
@@ -0,0 +1,122 @@
+    }
+  }
+
+  private static String resolveName(UUID uuid) throws IOException {
+    HttpURLConnection con =
+        (HttpURLConnection)
+            new URL("https://api.ashcon.app/mojang/v2/user/" + uuid.toString()).openConnection();
+    con.setRequestMethod("GET");
+    con.setRequestProperty("Accept", "application/json");
+    con.setConnectTimeout(5000);
+    con.setReadTimeout(5000);
+    BufferedReader br =
+        new BufferedReader(new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8));
+    StringBuilder response = new StringBuilder();
+    String line;
+    while ((line = br.readLine()) != null) {
+      response.append(line.trim());
+    }
+    br.close();
+    return GSON.fromJson(response.toString(), JsonObject.class).get("username").getAsString();
+  }
+
+  public static void resolveAll() {
+    Bukkit.getScheduler()
+        .runTaskAsynchronously(
+            PGM.get(),
+            () -> {
+              for (int i = 0; i < unresolved.size(); i++) {
+                try {
+                  UUID uuid = unresolved.get(i);
+                  cache.put(
+                      uuid,
+                      new CachedPlayer(
+                          uuid, NameCacheUtil.resolveName(uuid), System.currentTimeMillis()));
+                } catch (IOException ignored) {
+                }
+              }
+              unresolved.clear();
+              try {
+                writeCacheToDisk();
+              } catch (IOException ignored) {
+              }
+            });
+  }
+
+  public static void writeCacheToDisk() throws IOException {
+    JsonWriter writer =
+        new JsonWriter(new FileWriter(new File(PGM.get().getDataFolder(), "uuidcache.json")));
+    writer.beginArray();
+    for (Entry<UUID, CachedPlayer> entry : cache.entrySet()) {
+      writer.beginObject();
+      writer.name("uuid").value(entry.getKey().toString());
+      writer.name("name").value(entry.getValue().getName());
+      writer.name("timestamp").value(entry.getValue().getTimestamp());
+      writer.endObject();
+    }
+    writer.endArray();
+    writer.close();
+  }
+
+  public static void readCacheFromDisk() throws IOException {
+    BufferedReader bufferedReader =
+        new BufferedReader(new FileReader(new File(PGM.get().getDataFolder(), "uuidcache.json")));
+
+    JsonArray array = GSON.fromJson(bufferedReader, JsonArray.class);
+    for (int i = 0; i < array.size(); i++) {
+      JsonObject obj = array.get(i).getAsJsonObject();
+      CachedPlayer player =
+          new CachedPlayer(
+              UUID.fromString(obj.get("uuid").getAsString()),
+              obj.get("name").getAsString(),
+              obj.get("timestamp").getAsLong());
+      // If the player's name has been fetched within the last seven days
+      if (System.currentTimeMillis() - player.getTimestamp() < 604800000L) {
+        cache.put(player.getUUID(), player);
+      } else {
+        unresolved.add(player.getUUID());
+      }
+    }
```

Comment:
Same applies here, you can directly load the cache from GSON.
```java
    try (FileReader reader = new FileReader(new File(...))) {
        cache = GSON.fromJson(reader, new TypeToken<HashMap<UUID, NameCacheEntry>>(){}.getType());
    }
```

Also, you shouldn't avoid old names getting in the cache (or remove them from the cache), you can just both keep them, and add to unresolved to have them updated. And try to give `604800000L` a better name over a comment
```java
    long ONE_WEEK_AGO = System.currentTimeMillis() - (7 * 24 * 60 * 60 * 1000);
    cache.values().stream().filter(ce -> ce.timestamp < ONE_WEEK_AGO).forEach(ce -> unresolved.add(ce.uuid));
```